### ISSUE TITLE
Updating default wasm features

### ISSUE THREAD

**Issue by dschuff** [emscripten-core/emscripten#15376]
Do we have a policy on when we update the default set of wasm features enabled? I feel like we've discussed this a while ago but can't find any issues about it.

Here's the current state of things as best I can find.
Safari [shipped](https://developer.apple.com/documentation/safari-release-notes/safari-14_1-release-notes) the following features in version 14.1, which is included with MacOS Big Sur 11.3 and iOS 14.5 (both in late April 2021): Bigint, signext, and atomics (but not SharedArrayBuffer AFAIK, and not bulk memory)

Safari [15](https://developer.apple.com/documentation/safari-release-notes/safari-15-release-notes), included in MacOS Monterey 12, iOS 15 (Sept 20, 2021): Streaming compilation, bulk memory, reference types, nontrapping float-to-int.

Safari 15 shipped to [MacOS](https://9to5mac.com/2021/09/21/safari-15-with-new-tab-design-now-available-for-macos-big-sur-and-catalina-users/) versions older than the new MacOS 12, but I don't think that's true of iOS. (But I think iOS users update their OS faster than macOS users).

Firefox and Chrome have had all of those features for quite a while (other than reference types for some reason, which isn't too relevant for Emscripten).

Of the features Safari shipped in May, IMO really only BigInt is very valuable on its own.
Of the features shipped in September, nontrapping float-to-int and bulk memory will save some code size (and maybe performance, in the case of bulk memory).

How long do we want to wait before enabling these by default (and/or how many users do we want to catch?)

According to my best guess from [here](https://gs.statcounter.com/browser-version-market-share) desktop Safari 14.1 is currently about 72% of MacOS users, and my best guess from [here](https://gs.statcounter.com/ios-version-market-share/) is that iOS 14.5+ is currently about 67% of iOS users.

**Comment by tlively** [comment]
To be clear, we have complete control over both the LLVM and Emscripten defaults, and they can be entirely independent.

We previously decided to be somewhat aggressive about updating LLVM defaults (default after phase 4 + 6 months, IIRC) but actually executing on that decision fell through the cracks. Being more conservative in Emscripten makes sense to me, but we do want to have some concrete updating policy so that users do eventually get the benefits of new features.

**Comment by sbc100** [comment]
I think the in past @kripken has argued that we should keep llvm and emscripten in sync with defaults.

**Comment by dschuff** [comment]
I looked at the data from https://gs.statcounter.com/ios-version-market-share/mobile-tablet/worldwide/#monthly-202212-202312 (the graph isn't very useful because it doesn't aggregate by major version, but I downloaded the CSV) and found that the current market share for iOS v17 at 54.2%, v16 at 30.6%, v15 at 9.2, and everything lower at 6.0%. 
However the data from caniuse.com (see e.g. https://caniuse.com/mdn-javascript_operators_await_top_level for another feature which shipped in Safari 15) seems to suggest that iOS Safari versions 14.8 or less have only 0.85% share; that's a pretty big discrepancy, not sure what to make of that.

But either way it's maybe getting close to the point where we can enable bulk memory, nontrapping-fptoint, and BigInt by default.

**Comment by sbc100** [comment]
Lets do it!

**Comment by dschuff** [comment]
I think in order to flip the switch on this the way we did with signext, we'd want Binaryen bulk-memory and nontrapping-fptoint-lowering passes, right?

**Comment by sbc100** [comment]
Ah yes, that is correct.  Although I think only the bulkmemory ops that llvm itself generates would need to be in the lowering pass (i.e probably just memory.copy)

**Comment by sbc100** [comment]
Can we close this now or are we still waiting on some kind of official policy document?

**Comment by RReverser** [comment]
I think having an official policy for both Wasm and untranspiled JS features would still be helpful.

Something that is new since this issue had been raised - there are now official Baseline policies which all major browsers have agreed on implementing, so now users only need to target a specific Baseline year instead of thinking of individual browser versions.

You can see how that works for eg new Wasm features here: https://web.dev/blog/wasmgc-wasm-tail-call-optimizations-baseline cc @tomayac 

So Emscripten / LLVM would just need to document that it's targeting a specific baseline year, and bump it up from time to time.

**Comment by sbc100** [comment]
> So Emscripten / LLVM would just need to document that it's targeting a specific baseline year, and bump it up from time to time.

If our policy can be that simple that would be great. 

In practice, I imagine it could be more complicated because market shared doesn't always move in sync with number of years we go back.    i.e. targeting a baseline of -5 years might result in 3% of users not being able to run our code one year but 10% another year.   i.e. should we base our policy of number of years of baseline or one specific market share numbers? 

Obviously market share numbers are a complex beast and can vary depending on how/who you ask so we can avoid using those that would be great.

Honestly our current defacto policy seems to be working pretty well:  Bump the min versions whenever we want to move the codebase forward and we can have a general feeling that that impact will be minimal :)

**Comment by tomayac** [comment]
~Wasm compatibility also needs to cover more than browsers (which Baseline doesn't), for example, see the missing support of the tail call feature for, say, Wasmer. This is why I've created [`<wasm-compat>`](https://webassembly.org/news/2025-09-17-wasm-compat/#:~:text=wasm2c-,Tail%20Call%20(Phase%205),-Chrome) as a separate web component distinct from the Baseline one.~

**Comment by dschuff** [comment]
I don't think we want to take non-JS engine support into account for Emscripten though. Not that we want to leave anything behind on purpose but it's complicated enough as it is, and Emscripten has always unapologetically targeted JS engines first and foremost, and I don't think we want to hold anything back for missing support in non-JS engines.

**Comment by tomayac** [comment]
That's a fair decision to make, and one that's actually right in the [mission statement](https://emscripten.org/index.html#:~:text=Emscripten%20is%20a%20complete%20compiler%20toolchain%20to%20WebAssembly%2C%20using%20LLVM%2C%20with%20a%20special%20focus%20on%20speed%2C%20size%2C%20and%20the%20Web%20platform). Scratching my comment.

**Comment by dschuff** [comment]
We do get requests for support in targeting non-web JS environments such as node, Electron, and weird web environments like webviews, old environments like TVs, etc. Those are not the top priorities but we do often support those use cases with things like extra configurations, or best-effort support for different environments, etc because they are closer to the "real" web and a reasonable amount of effort. It probably wouldn't affect what we target by default though.

**Comment by juj** [comment]
> Emscripten has always unapologetically targeted JS engines first and foremost

The reason for this is that when compiling to JS->asm.js->wasm was created, no other Wasm VMs like Wasmer existed.

Then when those Wasm VMs came around, they mainly wanted to target building from LLVM/Clang directly to WebAssembly as a feature, rather than integrate with Emscripten to add support for targeting new runtimes. This make sense, since many of those runtimes do not host JavaScript at all, so large parts of "what Emscripten is all about" will be redundant to them.

So that is how these ended up being two separate worlds.

There are two flags, `-sSTANDALONE_WASM` and `-sPURE_WASI` currently that allow targeting Wasm-only VMs. But they might not be as full-featured as users of those VMs might need. This will be up to the individual contributors/champions to progress forward as usual.

----------

In general Emscripten already has a robust and fine-grained feature/capability tracking (the -sMIN_x_VERSION system and code in feature_matrix.py), so the "Baseline <year>" mechanism would be maybe a bit redundant for the project.

**Comment by RReverser** [comment]
> In general Emscripten already has a robust and fine-grained feature/capability tracking (the -sMIN_x_VERSION system and code in feature_matrix.py), so the "Baseline " mechanism would be maybe a bit redundant for the project.

The original intent was rather to specify the policy for defaults for those `-sMIN_x_VERSION` and when do we bump them, so these are slightly orthogonal.
